modules/jolt_physics/objects/jolt_soft_body_3d.cpp

```diff
@@ -424,6 +424,82 @@ bool JoltSoftBody3D::is_sleeping() const {
 	return !body->IsActive();
 }
 
+void JoltSoftBody3D::apply_vertex_impulse(int p_index, const Vector3 &p_impulse) {
+	ERR_FAIL_COND_MSG(!in_space(), vformat("Failed to apply impulse to '%s'. Doing so without a physics space is not supported when using Jolt Physics. If this relates to a node, try adding the node to a scene tree first.", to_string()));
+
+	ERR_FAIL_NULL(shared);
+	ERR_FAIL_INDEX(p_index, (int)shared->mesh_to_physics.size());
+	const size_t physics_index = (size_t)shared->mesh_to_physics[p_index];
+	if (pinned_vertices.has(physics_index)) {
+		ERR_FAIL_MSG("Cannot apply impulse or force on pinned point.");
+		return;
+	}
+
+	const float last_step = space->get_last_step();
+	if (unlikely(last_step == 0.0f)) {
+		return;
+	}
+
+	JoltWritableBody3D body = space->write_body(jolt_id);
+	ERR_FAIL_COND(body.is_invalid());
+
+	JPH::SoftBodyMotionProperties &motion_properties = static_cast<JPH::SoftBodyMotionProperties &>(*body->GetMotionPropertiesUnchecked());
+
+	JPH::Array<JPH::SoftBodyVertex> &physics_vertices = motion_properties.GetVertices();
+	JPH::SoftBodyVertex &physics_vertex = physics_vertices[physics_index];
+
+	physics_vertex.mVelocity += to_jolt(p_impulse) * physics_vertex.mInvMass;
+
+	wake_up();
+}
+
+void JoltSoftBody3D::apply_vertex_force(int p_index, const Vector3 &p_force) {
+	ERR_FAIL_COND_MSG(!in_space(), vformat("Failed to apply force to '%s'. Doing so without a physics space is not supported when using Jolt Physics. If this relates to a node, try adding the node to a scene tree first.", to_string()));
+
+	apply_vertex_impulse(p_index, p_force * space->get_last_step());
+}
+
+void JoltSoftBody3D::apply_central_impulse(const Vector3 &p_impulse) {
+	ERR_FAIL_COND_MSG(!in_space(), vformat("Failed to apply central impulse to '%s'. Doing so without a physics space is not supported when using Jolt Physics. If this relates to a node, try adding the node to a scene tree first.", to_string()));
+	ERR_FAIL_NULL(shared);
+
+	const float last_step = space->get_last_step();
+	if (unlikely(last_step == 0.0f)) {
+		return;
+	}
+
+	JoltWritableBody3D body = space->write_body(jolt_id);
+	ERR_FAIL_COND(body.is_invalid());
+
+	JPH::SoftBodyMotionProperties &motion_properties = static_cast<JPH::SoftBodyMotionProperties &>(*body->GetMotionPropertiesUnchecked());
+	JPH::Array<JPH::SoftBodyVertex> &physics_vertices = motion_properties.GetVertices();
+
+	const int mesh_vertex_count = shared->mesh_to_physics.size();
+
+	const JPH::Vec3 impulse = to_jolt(p_impulse) / mesh_vertex_count;
+
+	for (int i = 0; i < mesh_vertex_count; ++i) {
+		const size_t physics_index = (size_t)shared->mesh_to_physics[i];
+		if (pinned_vertices.has(physics_index)) {
+			continue;
+		}
+
+		JPH::SoftBodyVertex &physics_vertex = physics_vertices[physics_index];
+		physics_vertex.mVelocity += impulse * physics_vertex.mInvMass;
+	}
```

Comment:
This would make the loop a bit faster:

```suggestion
	for (JPH::SoftBodyVertex &physics_vertex : physics_vertices) {
	    if (physics_vertex.mInvMass > 0.0f) {
		    physics_vertex.mVelocity += impulse * physics_vertex.mInvMass;
		}
	}
```

Reply:
I think this change didn't work correctly.

https://github.com/user-attachments/assets/19ed2393-1af2-4c1a-965d-cc58466d8ebd

Reply:
I'll check what's going on.